urid/src/feature.rs

```diff
@@ -0,0 +1,124 @@
+}
+
+impl<'a> Feature<'a> for Map<'a> {
+    type RawDataType = sys::LV2_URID_Map;
+
+    fn from_raw_data(data: Option<&'a mut sys::LV2_URID_Map>) -> Option<Self> {
+        if let Some(internal) = data {
+            Some(Self { internal })
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a> Map<'a> {
+    /// Return the URID of the given URI.
+    ///
+    /// This method capsules the raw mapping method provided by the host. Therefore, it may not be very fast or even capable of running in a real-time environment. Instead of calling this method every time you need a URID, you should call it once and cache it.
+    pub fn map(&self, uri: &CStr) -> URID {
+        let handle = self.internal.handle;
+        let uri = uri.as_ptr();
+        unsafe { (self.internal.map.unwrap())(handle, uri) }
+    }
+}
+
+/// Host feature to revert the URI -> URID mapping.
+pub struct Unmap<'a> {
+    internal: &'a sys::LV2_URID_Unmap,
+}
+
+unsafe impl<'a> UriBound for Unmap<'a> {
+    const URI: &'static [u8] = sys::LV2_URID_UNMAP_URI;
+}
+
+impl<'a> Feature<'a> for Unmap<'a> {
+    type RawDataType = sys::LV2_URID_Unmap;
+
+    fn from_raw_data(data: Option<&'a mut sys::LV2_URID_Unmap>) -> Option<Self> {
+        if let Some(internal) = data {
+            Some(Self { internal })
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a> Unmap<'a> {
+    /// Return the URI of the given URID.
+    ///
+    /// This method capsules the raw mapping method provided by the host. Therefore, it may not be very fast or even capable of running in a real-time environment. Instead of calling this method every time you need a URID, you should call it once and cache it.
+    pub fn unmap(&self, urid: URID) -> &CStr {
+        let handle = self.internal.handle;
+        unsafe {
+            let uri = (self.internal.unmap.unwrap())(handle, urid);
+            CStr::from_ptr(uri)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::feature::*;
+    use std::collections::HashMap;
+    use std::ffi::{c_void, CStr};
+
+    unsafe extern "C" fn internal_mapping_fn(handle: *mut c_void, uri: *const i8) -> URID {
+        let handle = (handle as *mut HashMap<&CStr, URID>).as_mut().unwrap();
+        let uri = CStr::from_ptr(uri);
+        if !handle.contains_key(uri) {
+            handle.insert(uri, handle.len() as u32);
+        }
+        handle[uri]
+    }
+
+    unsafe extern "C" fn internal_unmapping_fn(handle: *mut c_void, urid: URID) -> *const i8 {
+        let handle = (handle as *mut HashMap<&CStr, URID>).as_mut().unwrap();
+        for key in handle.keys() {
+            if handle[key] == urid {
+                return key.as_ptr();
+            }
+        }
+        std::ptr::null()
+    }
+
+    #[test]
+    fn test_map_unmap() {
+        let mut internal_map: HashMap<&CStr, URID> = HashMap::new();
+        let mut sys_map = sys::LV2_URID_Map {
+            handle: (&mut internal_map) as *mut _ as *mut c_void,
```

Comment:
I know this is just a test, but I believe this here is undefined behavior: you have two active structs holding mutable references to the HashMap. I think the `Map` and `Unmap` structs here should be tied to a lifetime that to the Hashmap's (or to whatever is used to implement this).
Also, because in practice several (dozens of) plugins will each hold a reference to these features, the Map and Unmap features should be able to properly operate on `&T` instead of `&mut T` all the way. I am also not sure about which threading classes apply to these (this seems to be undocumented), so perhaps it needs to be `Sync` as well?

Reply:
The thing is: As you can see in [the docs](http://lv2plug.in/doc/html/group__urid.html), every map and unmap feature struct holds a **mutable** pointer to the handle and therefore, we have to provide something mutable too. We can not do much about this, but I just recognized that you can make this safer by hiding it behind a mutex! I will add that later.

Reply:
I was also thinking about using `Mutex`, however I think it should be left to the host to implement. We don't want to add such a cost to the host if not needed, i.e. a `RefCell` would be sufficient if the host is single-threaded (or in small tests like these, not that the overhead matters anyway though).

Reply:
(Nevermind this comment, did not see the commits you added there, my bad!)